Stabilize some Result methods as const

CDirkx opened the pull request:
Stabilize the following methods of Result as const:
 - `is_ok`
 - `is_err`
 - `as_ref`

A test is also included, analogous to the test for `const_option`.

These methods are currently const under the unstable feature `const_result` (tracking issue: #67520).
I believe these methods to be eligible for stabilization because of the stabilization of #49146 (Allow if and match in constants) and the trivial implementations, see also: [PR#75463](https://github.com/rust-lang/rust/pull/75463) and [PR#76135](https://github.com/rust-lang/rust/pull/76135).

Note: these methods are the only methods currently under the `const_result` feature, thus this PR results in the removal of the feature.

Related: #76225

Commit:
Stabilize some Result methods as const

Stabilize the following methods of `Result` as const:
 - `is_ok`
 - `is_err`
 - `as_ref`

Possible because of stabilization of #49146 (Allow if and match in constants).

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

CDirkx commented:
r? @ecstatic-morse

ecstatic-morse commented:
r? @BurntSushi Needs FCP

ecstatic-morse commented:
r? @KodrAus (Needs FCP)

CDirkx commented:
Created an issue for similar stabilizations: #76225

CDirkx committed:
Move const tests for `Result` to `library\core`

Part of #76268

KodrAus commented:
For reference, there's also an equivalent push to constify some `Result` methods at #76135

@rfcbot fcp merge

KodrAus commented:
@rfcbot fcp merge

josephlr commented:
There are some other methods we should also make const (either in this PR or a followup). They all have trivial match implementations and doesn't use any trait bounds:

[`Result::ok()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.ok)
[`Result::err()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.err)
[`Result::iter()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.iter)
[`Result::and()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.and)
[`Result::or()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.or)
[`Result::unwrap_or()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.unwrap_or)
[`Result::transpose()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.transpose)

[`Result::flatten()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.flatten) also has a trivial impl, but that's unstable anyway (see https://github.com/rust-lang/rust/issues/70142#issuecomment-691453040)

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 787b2707a77fa1a98f512a4905662a1a16daf13c has been approved by `dtolnay`

<!-- @bors r=dtolnay 787b2707a77fa1a98f512a4905662a1a16daf13c -->
<!-- homu: {"type":"Approved","sha":"787b2707a77fa1a98f512a4905662a1a16daf13c","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit 787b2707a77fa1a98f512a4905662a1a16daf13c with merge a51855f645eb6b86d1c6eb5656ae3394c9949a05...
<!-- homu: {"type":"BuildStarted","head_sha":"787b2707a77fa1a98f512a4905662a1a16daf13c","merge_sha":"a51855f645eb6b86d1c6eb5656ae3394c9949a05"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1127239624)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1127239624","builder_name":"checks-actions"} -->

ecstatic-morse commented:
@CDirkx Test failures are legit. Clippy has a lint that suggests `is_err()` instead of `if let Err(_)`, but it's disabled in `const fn`. With this PR, that exception is no longer necessary and the test output can be updated.

CDirkx committed:
Update Clippy testcases

Update the test `redundant_pattern_matching`: check if `is_ok` and `is_err` are suggested within const contexts.
Also removes the `redundant_pattern_matching_const_result` test, as it is no longer needed.

CDirkx commented:
I updated the testcases, when this get's merged I'll update the testcases using `Option` as well and sync these changes to the Clippy repo.

RalfJung commented:
How can it be enough to update the testcases, doesn't the clippy code emitting these lints also need to be updated?

ebroto commented:
Clippy is checking if the suggested methods are `const fn` in const contexts, so it works without changing the logic because they are after this change.

But once the `Option` methods are stabilized as `const fn` the function `can_suggest` and calls to it should be removed as there would be no special cases left.

CDirkx commented:
That's what I thought as well (https://github.com/rust-lang/rust/pull/76135#issuecomment-695358883), but the implementation of `can_suggest` is not specific to `is_err` etc., and thus could be used in the future if Clippy were to ever suggest other unstable const methods. 

https://github.com/rust-lang/rust/blob/a3bc0e752fad96f537b73f4e9bc805a73d404f7b/src/tools/clippy/clippy_lints/src/matches.rs#L1649-L1670 

However it would be unused now, so indeed maybe better to remove it.

CDirkx committed:
Remove `can_suggest` check for `is_ok` and `is_err`.

`is_ok` and `is_err` are stabilized as const and can thus always be suggested.

RalfJung commented on src/tools/clippy/clippy_lints/src/matches.rs:
```diff
@@ -1562,8 +1562,8 @@ mod redundant_pattern_match {
-                            || can_suggest(cx, hir_id, sym!(result_type), "is_ok"),
-                            || can_suggest(cx, hir_id, sym!(result_type), "is_err"),
+                            || true,
+                            || true
```
```suggestion
                            || true,
```

CDirkx committed:
Update src/tools/clippy/clippy_lints/src/matches.rs

Co-authored-by: Ralf Jung <post@ralfj.de>

RalfJung commented:
@bors r=dtolnay

bors commented:
:pushpin: Commit bf70e21a7e48008a8c1d82c3b0376e574f9bc91b has been approved by `dtolnay`

<!-- @bors r=dtolnay bf70e21a7e48008a8c1d82c3b0376e574f9bc91b -->
<!-- homu: {"type":"Approved","sha":"bf70e21a7e48008a8c1d82c3b0376e574f9bc91b","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit bf70e21a7e48008a8c1d82c3b0376e574f9bc91b with merge b873fa6d42cf305131d2583d03b84686e5e40f2e...
<!-- homu: {"type":"BuildStarted","head_sha":"bf70e21a7e48008a8c1d82c3b0376e574f9bc91b","merge_sha":"b873fa6d42cf305131d2583d03b84686e5e40f2e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1140556883), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36729)
Approved by: dtolnay
Pushing b873fa6d42cf305131d2583d03b84686e5e40f2e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36729","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1140556883"},"merge_sha":"b873fa6d42cf305131d2583d03b84686e5e40f2e"} -->
